### PR TITLE
Multi agent reinforcement training demo

### PR DESCRIPTION
**Pull request by JingfeiPeng**
# Game of Tag
This directory contains a a multi-agent adversarial training demo. In the demo, there is a predator vehicle and a prey vehicle.
The predator vehicle's goal is to catch the prey, and the prey vehicle's goal is to avoid getting caught. 

## Run training
python examples/game_of_tag/game_of_tag.py examples/game_of_tag/scenarios/game_of_tag_demo_map/

## Run checkpoint
python examples/game_of_tag/run_checkpoint.py examples/game_of_tag/scenarios/game_of_tag_demo_map/

### Rewards
The formula for reward is 0.5/(distance-COLLIDE_DISTANCE)^2 and capped at 10

- COLLIDE_DISTANCE is the observed distance when two vehicle collides. Since the position of two vehicle is at the center, the distance when collesion happens is not exactly 0. 

### Common Reward:
    Off road: -10

#### Prey:
    collide with predator: -10
    distance to predator(d): 0.5/(d-COLLIDE_DISTANCE)^2 
#### Predator:
    collide with predator: -10
    distance to predator(d): 0.5/(d-COLLIDE_DISTANCE)^2 

### Action:
4 speed modes: [0, 3, 6, 9]

lanechange: -1, 0, 1

## Output a model:
Currently Rllib does not have implementation for exporting a pytorch model. 

Replace `export_model`'s implementation in `ray/rllib/policy/torch_policy.py` to the following:
```
torch.save(self.model.state_dict(),f"{export_dir}/model.pt")
```
Then follow the steps in game_of_tag.py to export the model.

## Possible next steps
- increase the number of predators and preys to 2 predators and 2 preys. 
This requires modeling the reward to still be a zero sum game which could be difficult when 2 predators are close to
1 prey, causing the sum of predator and prey rewards to no longer be 0. 
- add an attribute in observations to display whether the ego car is in front of the target vehicle or behind it, this helps
to let ego vehicle know whether it should slow down or speed up

### CODE REVIEW

File: smarts/env/rllib_hiway_env.py [hunk 1]
```diff
@@ -147,7 +149,13 @@ def step(self, agent_actions):
 
         for done in dones.values():
             self._dones_registered += 1 if done else 0
-        dones["__all__"] = self._dones_registered == len(self._agent_specs)
+        # TODO temporary solution for game of tag: stop the episode when 1 vehicle is done
+        # so that the other vehicle does not train when the opponent is not present, which
+        # causes noisy in training
+        dones["__all__"] = self._dones_registered > 0
```
Comment (JingfeiPeng) [inline smarts/env/rllib_hiway_env.py:155]: Will create a PR on adding configurable DoneCreteria for stopping simulation when certain vehicles are done, and wiill remove this if that PR can be merged

File: examples/game_of_tag/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+#### Prey:
+    collide with predator: -10
+    distance to predator(d): 0.5/(d-COLLIDE_DISTANCE)^2 
+#### Predator:
+    collide with predator: -10
+    distance to predator(d): 0.5/(d-COLLIDE_DISTANCE)^2 
+
+### Action:
+4 speed modes: [0, 3, 6, 9]
```
Comment (Gamenot) [inline examples/game_of_tag/README.md:28]:
```suggestion
Speed modes: [0, 3, 6, 9]
```

File: examples/game_of_tag/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+### Common Reward:
+    Off road: -10
+
+#### Prey:
+    collide with predator: -10
+    distance to predator(d): 0.5/(d-COLLIDE_DISTANCE)^2 
+#### Predator:
+    collide with predator: -10
+    distance to predator(d): 0.5/(d-COLLIDE_DISTANCE)^2
```
Comment (Gamenot) [inline examples/game_of_tag/README.md:25]:
```suggestion
#### Prey:
    Collision with predator: -10
    Distance to predator(d): 0.5/(d-COLLIDE_DISTANCE)^2 
#### Predator:
    Collision with predator: -10
    Distance to predator(d): 0.5/(d-COLLIDE_DISTANCE)^2 
```

File: examples/game_of_tag/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+torch.save(self.model.state_dict(),f"{export_dir}/model.pt")
+```
+Then follow the steps in game_of_tag.py to export the model.
+
+## Possible next steps
+- increase the number of predators and preys to 2 predators and 2 preys. 
+This requires modeling the reward to still be a zero sum game which could be difficult when 2 predators are close to
+1 prey, causing the sum of predator and prey rewards to no longer be 0. 
+- add an attribute in observations to display whether the ego car is in front of the target vehicle or behind it, this helps
+to let ego vehicle know whether it should slow down or speed up
```
Comment (Gamenot) [inline examples/game_of_tag/README.md:46]:
```suggestion
## Possible next steps
- Increase the number of agents to 2 predators and 2 prey. 
This requires modelling the reward to still be a zero sum game. The complication can be understood from 
how to model the distance reward between 2 predators and 1 prey. If the reward is only from nearest predator 
to nearest prey, the sum of predator and prey rewards will no longer be 0 because 2 predators will be getting full 
reward from 1 prey. This will require the predators to know about each other or the prey to know about other prey, 
and the prey to know about multiple predators.
- Add an attribute in observations to display whether the ego car is in front of the target vehicle or behind it, this may 
help to let ego vehicle know whether it should slow down or speed up
```
